---
title: "R Notebook"
output: html_notebook
---

## Setup

```{r, warning=FALSE, error=FALSE}
rm(list = ls())

library(tidyverse)
library(ggplot2)
library(ggthemes)
library(glue)
library(RIdeogram)
library(openxlsx)
library(latex2exp)
library(cowplot)
library(magick)

theme_solarized_2() |>
  theme_set()

ensure.output <- function(directory) {
  output_dir <- "output"
  output_sub_dir <- "{output_dir}/{directory}" |> glue()
  output_sub_dir |> dir.create(showWarnings = FALSE, recursive = TRUE)
  output_sub_dir
}
```

## CQ

### Load datasets

```{r}
read.single_alignment <- function(file) {
  url_prefix <- "http://imlspenticton.uzh.ch/dump/ece/gDNA/CQ/Depths"
  "{url_prefix}/{file}" |>
    glue() |>
    read_tsv(show_col_types = FALSE) |>
    rename(c(
      "contig" = "#rname",
      "length" = "endpos",
      "depth" = "meandepth")) |>
    select(contig, length, coverage, depth)
}

read.alignment <- function(file.female, file.male) {
  alignment.female <- read.single_alignment(file.female)
  alignment.male <- read.single_alignment(file.male)
  
  alignment.female |>
    left_join(
      alignment.male, 
      by = c("contig", "length"),
      suffix = c(".female", ".male")) |>
    mutate(cq = depth.female / depth.male)
}

m1.alignment <- read.alignment(
  file.female = "contigScale/MIaabys_female_on_MIcontigs_depth_perContig.txt",
  file.male = "contigScale/MIaabys_male_on_MIcontigs_depth_perContig.txt") |>
  filter(depth.male >= 1.0) 

"M1:" |> print()
m1.alignment |> summary() 


control.alignment <- read.alignment(
  file.female = "controlGenes/MIaabys_female_on_controls.txt",
  file.male = "controlGenes/MIaabys_male_on_controls.txt")

"Control:" |> print()
control.alignment |> summary() 
```

### Normalize them

```{r}
m1.alignment <- m1.alignment |>
  mutate(cq.norm = cq / mean(control.alignment$cq))

m1.alignment$cq.norm |> summary()
```

### Load synteny

```{r}
synteny <- read.xlsx("http://imlspenticton.uzh.ch/dump/ece/cDNA/MachineLearning/PredictedData/MI_contigs_predicted.xlsx", sheet = 1) |>
  select(contig, allocation)
m1.alignment <- m1.alignment |>
  left_join(synteny, by = "contig")
```

### Plot

```{r}
cutoff <- 0.2

plot.cq <- function(data, cutoff, title) {
  cutoff.count <- data |>
    filter(cq.norm <= cutoff) |>
    nrow()

  data |>
    ggplot(aes(x = log(length), y = log(cq.norm))) +
    labs(
      title = title,
      x = "ln length",
      y = "ln CQ") +
    geom_point() +
    annotate(
      "text", 
      x = 15, 
      y = -3.5, 
      label = glue("{cutoff.count} contigs seem to be under the cutoff CQ <= {cutoff}")) +
    geom_hline(yintercept = log(cutoff), color = "salmon")
}

chromosomes <- c("2L", "NALL")
m1.alignment |> plot.cq(cutoff, "CQ, all contigs")
m1.alignment |>
  filter(allocation %in% chromosomes) |> 
  plot.cq(cutoff, "CQ, contigs mapping to {glue_collapse(chromosomes, sep = ', ')}" |> glue())
```

### Save data

```{r}
output_dir <- ensure.output("cq")

m1.alignment |>
  write_tsv("{output_dir}/cq_with_synteny.tsv" |> glue())
```

## Karyotype

### Load Synteny

```{r ProtoX search}
read.synteny <- function(file) {
  url_prefix <- "http://imlspenticton.uzh.ch/dump/ece/cDNA/BUSCO/OrthoMCL/SynthenyMaps"
  "{url_prefix}/{file}" |>
    glue() |>
    read_tsv(show_col_types = FALSE)
}

synteny.dmel_to_m5 <- read.synteny("Dmel_vs_MV/Dmel_MV_sinteny.txt")

bugz <- "tig00000533"
bugz.orthologs <- synteny.dmel_to_m5 |>
  filter(Species2 == bugz) |>
  select(Ortho) |>
  unique() |>
  unlist()       

synteny.dmel_to_m1 <- read.synteny("Dmel_vs_MI/Dmel_MI_sinteny.txt") |>
  filter(Ortho %in% bugz.orthologs)

synteny.dmel_to_m1 <- synteny.dmel_to_m1 |>
  count(Ortho) |>
  rename(c("Freq" = "n")) |>
  left_join(synteny.dmel_to_m1, by = "Ortho") |>
  filter(Freq <= 4)

synteny.dmel_to_m1 |> 
  count(Species2) |>
  arrange(n |> desc())
```

### Build Orthology of MI to MV

```{r}
# These have a good amount of distincly mapped genes
contigs_of_interest <- c(
  "tig00003344", 
  "tig00003381", 
  "tig00009311", 
  "tig00009367",
  "tig00009613",
  "tig00009652",
  "tig00009658", 
  "tig00009705", 
  "tig00009706", 
  "tig00009708", 
  "tig00009710")

synteny.dmel_to_m1 <- synteny.dmel_to_m1 |>
  filter(Species2 %in% contigs_of_interest)    

synteny.bugz_dmel_to_m5 <- synteny.dmel_to_m5 |>
  filter(Species2 == bugz) |>
  select(Ortho, Species2, Start2, End2) |>  
  rename(c(
    "Species1" = "Species2",
    "Start1" = "Start2",
    "End1" = "End2"
  )) 

synteny.bugz_m1_to_m5 <- synteny.dmel_to_m1 |>
  select(Ortho, Species2, Start2, End2) |>
  left_join(synteny.bugz_dmel_to_m5, by = "Ortho") |>
  select(Species1, Start1, End1, Species2, Start2, End2)
```

### Read Candidates
```{r}
candidates <- read_csv("data/candidates.csv") |>
  mutate(STATUS = STATUS |> factor())
```

### Compute Karyotype

```{r}
# The rest are at the end; that's ok
order_on_synteny <- c(
  'tig00003381', 
  'tig00003344',
  'tig00009652', 
  'tig00009658',
  'tig00009706', 
  'tig00009705',
  'tig00009367', 
  'tig00009311', 
  'tig00009710')

alignment.m1_male <- read.single_alignment("contigScale/MIaabys_male_on_MIcontigs_depth_perContig.txt") |>
  select(contig, coverage, length) |>
  filter(contig %in% synteny.dmel_to_m1$Species2) |>
  arrange(factor(contig, levels = order_on_synteny))

n <- 1 + synteny.dmel_to_m1$Species2 |>
  unique() |>
  length()

m5_contig_length <- 17343608
karyotype <- tibble(
  Chr = c(bugz, alignment.m1_male$contig),
  Start = replicate(n, 1) |> as.numeric(),
  End =  c(m5_contig_length, alignment.m1_male$length),
  fill =  c("95FF7D", replicate(n - 1, "8AFFBB")),
  species = c("ProtoX", replicate(n - 1 , "target")), # We will replace "target" later
  size = replicate(n, 12),
  color = replicate(n, 25252))
```

### Arrange chromosomes by CQ
```{r}
synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5 |>
  left_join(m1.alignment |> select(contig, cq.norm), by = c(
    "Species2" = "contig"
  )) |>
  mutate(chromosome = if_else(cq.norm < 0.4, "Y", "X")) |>
  select(-cq.norm)
```

### Use prior knowledge of M1 candidates
- contigs with candidates are assumed to be proto-Y
- the contig with SAM *must* be proto-Y
- on clash: the contig with the most candidates is assumed to be proto-Y
```{r}
candidates <- read_csv("candidates.csv") |>
  mutate(STATUS = STATUS |> as.factor(),
         `contig allocation` = `contig allocation` |> as.factor()) |>
  select(contig = tname, STATUS) |>
  group_by(contig) |>
  summarize(contig, candidate_count = n(), has_sam = "good candidate: SAM" %in% STATUS)

synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |> 
  mutate(chromosome = if_else(Species2 %in% candidates$contig, "Y", chromosome))
```

### Manually fix wrong double assignments based on CQ comparison
Some contigs have both the proto-Y[M1] and the proto-X[M1] assigned to the same chromosome. Check these pairs manually and move them.
```{r}
manual_x <- c(
  "tig00003381",
  "tig00009705",
  "tig00009613"
)
manual_y <- c(
  "tig00009652",
  "tig00009708" # This guy is Y even though CQ says otherwise
)

synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |> 
  mutate(chromosome = if_else(Species2 %in% manual_x, "X", chromosome)) |> 
  mutate(chromosome = if_else(Species2 %in% manual_y, "Y", chromosome))
```

### Flip inverted contigs
Chosen by eye
```{r}
to_flip <- c(
  # X
  "tig00003344",
  "tig00009658",
  "tig00009706",
  "tig00009705",
  "tig00009613",
  "tig00009708",
  # Y
  "tig00009652",
  "tig00009710"
)
synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |>
  left_join(m1.alignment |> select(contig, length), by = c(
    "Species2" = "contig"
  )) |>
  mutate(Start2.new = if_else(Species2 %in% to_flip, length - End2 , Start2),
         End2.new = if_else(Species2 %in% to_flip, length - Start2 , End2)) |>
  mutate(Start2 = Start2.new,
         End2 = End2.new) |>
  select(-length,
         -Start2.new,
         -End2.new)
synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |>
  mutate(fill = "FF6E63")
```

### Trim ends of MI contigs
```{r}
start_contigs <- c(
  "tig00003344", # Y
  "tig00003381" # X
)
end_contigs <- c(
  "tig00009613", # Y,
  "tig00009708" # X
)

synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |>
  group_by(Species2) |>
  summarize(
    Species1, Start1, End1, Species2, Start2, End2, chromosome, fill,
    # -1 because a Start of 1 means no offset
    Start2.offset = if_else(Species2 %in% start_contigs, 
                            min(Start2) - 1, 
                            0)) |>
  mutate(
    Start2 = Start2 - Start2.offset,
    End2 = End2 - Start2.offset) |>
  ungroup()

karyotype <- karyotype |>
  left_join(synteny.bugz_m1_to_m5.filtered |> select(Species2, Start2.offset, End2), by = c(
    "Chr" = "Species2"
  )) |>
  group_by(Chr) |>
  summarize(
    Start,
    End = if_else(Chr %in% end_contigs, max(End2), if_else(Chr != bugz, End - Start2.offset, End)),
    fill, species, size, color) |>
  distinct() |>
  ungroup() |>
  arrange(factor(Chr, levels = order_on_synteny))

synteny.bugz_m1_to_m5.filtered <- synteny.bugz_m1_to_m5.filtered |>
  # ideogram requires a specific column order
  select(Species1, Start1, End1, Species2, Start2, End2, chromosome, fill)
``` 

### Create And Save Ideogram

#### Functions

```{r}
get_rank <- function(to_rank, ranks) {
  for (i in ranks |> seq_along()) {
    contig <- ranks[[i]]
    to_rank <- replace(to_rank, to_rank == contig, i)
  }
  to_rank |> as.integer()
}

get_filtered_ideogram_data <- function(synteny, chrom) {
  karyotype.filtered <- karyotype |> 
    filter(species == "ProtoX") |>
    rbind(karyotype |> filter(Chr %in% synteny$Species2))

  karyotype.filtered <- karyotype.filtered |>
    mutate(species = if_else(species == "target", 
                             "Proto{chrom} M1" |> glue(), 
                             "ProtoX M5"))
  
  contig_head <- karyotype.filtered |>
    select(Chr) |>
    slice(1) |>
    unlist()
  
  contig_tail <- karyotype.filtered |>
    select(Chr) |>
    slice(-1) |>
    unlist()
  
  synteny.filtered <- synteny
  synteny.filtered$Species1 <- synteny$Species1 |>
    get_rank(contig_head)
  synteny.filtered$Species2 <- synteny$Species2 |>
    get_rank(contig_tail)
  list(
    karyotype = karyotype.filtered,
    synteny = synteny.filtered
  )
}

get_ideogram_path <- function(chrom) {
  output_dir <- ensure.output("karyotype")
  "{output_dir}/ideogram_{chrom}" |> glue()
}
create_ideogram <- function(chrom) {
  ideogram_path <- get_ideogram_path(chrom)
  data <- synteny.bugz_m1_to_m5.filtered |> 
    filter(chromosome == chrom) |> 
    select(-chromosome) |>
    get_filtered_ideogram_data(chrom)
  
  ideogram(
    karyotype = data$karyotype,
    synteny = data$synteny |> as.data.frame(),
    output = "{ideogram_path}.svg" |> glue())

  convertSVG(
    "{ideogram_path}.svg" |> glue(),
    device = "png", 
    file = "{ideogram_path}.png" |> glue(), 
    dpi = 500)
}
```

#### Creation

```{r}
create_ideogram("X")
create_ideogram("Y")
```

#### Plotting

```{r}
show_ideogram <- function(chrom) {
  ideogram_path <- get_ideogram_path(chrom)
  ggdraw() +
    draw_image("{ideogram_path}.png" |> glue())
}
#show_ideogram("X")
#show_ideogram("Y")
```

```{r}
karyotype |> 
  left_join(synteny.bugz_m1_to_m5.filtered |> select(Species2, chromosome) , by = c(
    "Chr" = "Species2"
  )) |> 
  distinct() |>
  rename(c("proto_chromosome" = "chromosome")) |>
  group_by(proto_chromosome) |>
  # Offset of + 1 because `Start = 1` means no shift
  summarize(proto_chromosome, length_mb = sum(End - Start + 1) / 10^6) |>
  distinct()
```


## Iterative Testing Ground

```{r}
to_check = c(
  # Y
  "tig00003381", # -> X
  "tig00003344", # 
  "tig00009705", # -> X
  "tig00009367", # SAM
  "tig00009710", # 
  "tig00009613", # -> X because of candidate list
  # X
  "tig00009652", # -> Y
  "tig00009658"
  #"tig00009311"
)

m1.alignment |>
  filter(contig %in% to_check) |>
  select(contig, cq.norm)
```

```{r}
candidates <- read_csv("candidates.csv") |>
  mutate(STATUS = STATUS |> as.factor(),
         `contig allocation` = `contig allocation` |> as.factor())

interest <- c(
  "tig00003381"
)
candidates |>
  filter(tname %in% interest) |>
  select(STATUS)
```
