---
title: "R Notebook"
output: html_notebook
---

## Setup

```{r, warning=FALSE, error=FALSE}
rm(list = ls())

library(tidyverse)
library(ggplot2)
library(ggthemes)
library(glue)
library(RIdeogram)
library(openxlsx)
library(latex2exp)
library(cowplot)
library(magick)

theme_solarized_2() |>
  theme_set()

ensure.output <- function(directory) {
  output_dir <- "output"
  output_sub_dir <- "{output_dir}/{directory}" |> glue()
  output_sub_dir |> dir.create(showWarnings = FALSE, recursive = TRUE)
  output_sub_dir
}
```

## CQ

### Load datasets

```{r}
read.single_alignment <- function(file) {
  url_prefix <- "http://imlspenticton.uzh.ch/dump/ece/gDNA/CQ/Depths"
  "{url_prefix}/{file}" |>
    glue() |>
    read_tsv(show_col_types = FALSE) |>
    rename(c(
      "contig" = "#rname",
      "length" = "endpos",
      "depth" = "meandepth")) |>
    select(contig, length, coverage, depth)
}

read.alignment <- function(file.female, file.male) {
  alignment.female <- read.single_alignment(file.female)
  alignment.male <- read.single_alignment(file.male)
  
  alignment.female |>
    left_join(
      alignment.male, 
      by = c("contig", "length"),
      suffix = c(".female", ".male")) |>
    mutate(cq = depth.female / depth.male)
}

m1.alignment <- read.alignment(
  file.female = "contigScale/MIaabys_female_on_MIcontigs_depth_perContig.txt",
  file.male = "contigScale/MIaabys_male_on_MIcontigs_depth_perContig.txt") |>
  filter(depth.male >= 1.0) 

"M1:" |> print()
m1.alignment |> summary() 


control.alignment <- read.alignment(
  file.female = "controlGenes/MIaabys_female_on_controls.txt",
  file.male = "controlGenes/MIaabys_male_on_controls.txt")

"Control:" |> print()
control.alignment |> summary() 
```

### Normalize them

```{r}
m1.alignment <- m1.alignment |>
  mutate(cq.norm = cq / mean(control.alignment$cq))

m1.alignment$cq.norm |> summary()
```

### Load synteny

```{r}
synteny <- read.xlsx("http://imlspenticton.uzh.ch/dump/ece/cDNA/MachineLearning/PredictedData/MI_contigs_predicted.xlsx", sheet = 1) |>
  select(contig, allocation)
m1.alignment <- m1.alignment |>
  left_join(synteny, by = "contig")
```

### Plot

```{r}
cutoff <- 0.2

plot.cq <- function(data, cutoff, title) {
  cutoff.count <- data |>
    filter(cq.norm <= cutoff) |>
    nrow()

  data |>
    ggplot(aes(x = log(length), y = log(cq.norm))) +
    labs(
      title = title,
      x = "ln length",
      y = "ln CQ") +
    geom_point() +
    annotate(
      "text", 
      x = 15, 
      y = -3.5, 
      label = glue("{cutoff.count} contigs seem to be under the cutoff CQ <= {cutoff}")) +
    geom_hline(yintercept = log(cutoff), color = "salmon")
}

chromosomes <- c("2L", "NALL")
m1.alignment |> plot.cq(cutoff, "CQ, all contigs")
m1.alignment |>
  filter(allocation %in% chromosomes) |> 
  plot.cq(cutoff, "CQ, contigs mapping to {glue_collapse(chromosomes, sep = ', ')}" |> glue())
```

### Save data

```{r}
output_dir <- ensure.output("cq")

m1.alignment |>
  write_tsv("{output_dir}/cq_with_synteny.tsv" |> glue())
```

## Karyotype

### Load Synteny

```{r ProtoX search}
read.synteny <- function(file) {
  url_prefix <- "http://imlspenticton.uzh.ch/dump/ece/cDNA/BUSCO/OrthoMCL/SynthenyMaps"
  "{url_prefix}/{file}" |>
    glue() |>
    read_tsv(show_col_types = FALSE)
}

synteny.dmel_to_m5 <- read.synteny("Dmel_vs_MV/Dmel_MV_sinteny.txt")

proto_x_m5 <- "tig00000533"
proto_x_m5.orthologs <- synteny.dmel_to_m5 |>
  filter(Species2 == proto_x_m5) |>
  select(Ortho) |>
  unique() |>
  unlist()       

synteny.dmel_to_m1 <- read.synteny("Dmel_vs_MI/Dmel_MI_sinteny.txt") |>
  filter(Ortho %in% proto_x_m5.orthologs)

synteny.dmel_to_m1 <- synteny.dmel_to_m1 |>
  count(Ortho) |>
  rename(c("Freq" = "n")) |>
  left_join(synteny.dmel_to_m1, by = "Ortho") |>
  filter(Freq <= 4)

synteny.dmel_to_m1 |> 
  count(Species2) |>
  arrange(n |> desc())
```

### Build Orthology of MI to MV

```{r}
# These have a good amount of distincly mapped genes
# The order is important for the ideogram. It was judged by eye.
contigs_of_interest <- c(
  'tig00003381', 
  'tig00003344',
  'tig00009652', 
  'tig00009658',
  'tig00009706', 
  'tig00009705',
  'tig00009367', 
  'tig00009311', 
  'tig00009710',
  'tig00009613',
  'tig00009708')

synteny.dmel_to_m1 <- synteny.dmel_to_m1 |>
  filter(Species2 %in% contigs_of_interest)    

synteny.proto_x_m5 <- synteny.dmel_to_m5 |>
  filter(Species2 == proto_x_m5) |>
  select(Ortho, Species2, Start2, End2) |>  
  rename(c(
    "Species1" = "Species2",
    "Start1" = "Start2",
    "End1" = "End2"
  )) 

synteny <- synteny.dmel_to_m1 |>
  select(Ortho, Species2, Start2, End2) |>
  left_join(synteny.proto_x_m5, by = "Ortho") |>
  select(Species1, Start1, End1, Species2, Start2, End2)
```

### Read Candidates
```{r}
candidates <- read_csv("data/candidates.csv") |>
  mutate(STATUS = STATUS |> factor())
```

### Compute Karyotype

```{r}
alignment.m1_male <- read.single_alignment("contigScale/MIaabys_male_on_MIcontigs_depth_perContig.txt") |>
  select(contig, coverage, length) |>
  filter(contig %in% synteny.dmel_to_m1$Species2) |>
  arrange(factor(contig, levels = contigs_of_interest))

n <- 1 + synteny.dmel_to_m1$Species2 |>
  unique() |>
  length()

proto_x_m5.length <- 17343608
karyotype <- tibble(
  Chr = c(proto_x_m5, alignment.m1_male$contig),
  Start = replicate(n, 1) |> as.numeric(),
  End =  c(proto_x_m5.length, alignment.m1_male$length),
  fill =  c("95FF7D", replicate(n - 1, "8AFFBB")),
  species = c("ProtoX", replicate(n - 1 , "target")), # We will replace "target" later
  size = replicate(n, 12),
  color = replicate(n, 25252))
```

### Arrange chromosomes by CQ
```{r}
synteny <- synteny |>
  left_join(m1.alignment |> select(contig, cq.norm), by = c(
    "Species2" = "contig"
  )) |>
  mutate(chromosome = if_else(cq.norm < 0.0, "Y", "X")) |>
  select(-cq.norm)
```

### Use prior knowledge of M1 candidates
- contigs with candidates are assumed to be proto-Y
- the contig with SAM *must* be proto-Y
- on clash: the contig with the most candidates is assumed to be proto-Y
```{r}
candidates <- read_csv("candidates.csv") |>
  mutate(STATUS = STATUS |> as.factor(),
         `contig allocation` = `contig allocation` |> as.factor()) |>
  select(contig = tname, status = STATUS, cq = `candidate CQ`) |>
  group_by(contig) |>
  summarize(
    contig, 
    candidate_count = n(), 
    has_sam = "good candidate: SAM" %in% status,
    cq.sum = sum(cq)) |>
  distinct()

synteny <- synteny |> 
  mutate(chromosome = if_else(Species2 %in% candidates$contig, "Y", chromosome))
```

### Fix clashes
```{r}
# 9652 X, 9658 Y
# Both significant candidate, no one knows
clashes_on_y <- synteny |> 
  distinct() |> 
  filter(chromosome == "Y") |> 
  group_by(Start1) |> 
  summarize(
    a = Species2 |> min(), 
    b = Species2 |> max()) |> 
  filter(a != b) |> 
  select(-Start1) |>
  distinct() |>
  left_join(candidates |> 
              select(
                contig,
                a.candidate_count = candidate_count, 
                a.has_sam = has_sam,
                a.cq.sum = cq.sum), 
            by = c("a" = "contig")) |>
  left_join(candidates |> 
              select(
                contig,
                b.candidate_count = candidate_count, 
                b.has_sam = has_sam,
                b.cq.sum = cq.sum), 
            by = c("b" = "contig")) |>
  distinct()
```

```{r}
for (i in 1:nrow(clashes_on_y)) {
  a <- clashes_on_y$a[i]
  b <- clashes_on_y$b[i]
  a.count <- clashes_on_y$a.candidate_count[i]
  b.count <- clashes_on_y$b.candidate_count[i]
  a.cq <- clashes_on_y$a.cq.sum[i]
  b.cq <- clashes_on_y$b.cq.sum[i]
  
  a.is_y <- (a.count > b.count) | clashes_on_y$a.has_sam[i]
  b.is_y <- (a.count < b.count) | clashes_on_y$b.has_sam[i]
  if (!(a.is_y | b.is_y)) {
    a.is_y <- a.cq < b.cq
  }
  if (a.is_y) {
    synteny <- synteny |> 
      mutate(chromosome = if_else(Species2 == b, "X", chromosome))
  } else {
    synteny <- synteny |> 
      mutate(chromosome = if_else(Species2 == a, "X", chromosome))
  }
}
```


### Flip inverted contigs
Chosen by eye
```{r}
to_flip <- c(
  # X
  "tig00003344",
  "tig00009658",
  "tig00009706",
  "tig00009705",
  "tig00009613",
  "tig00009708",
  # Y
  "tig00009652",
  "tig00009710"
)
synteny <- synteny |>
  left_join(m1.alignment |> select(contig, length), by = c(
    "Species2" = "contig"
  )) |>
  mutate(Start2.new = if_else(Species2 %in% to_flip, length - End2 , Start2),
         End2.new = if_else(Species2 %in% to_flip, length - Start2 , End2)) |>
  mutate(Start2 = Start2.new,
         End2 = End2.new) |>
  select(-length,
         -Start2.new,
         -End2.new)
synteny <- synteny |>
  mutate(fill = "FF6E63")
```

### Trim ends of MI contigs
```{r}
karyotype.length <- karyotype |> nrow()
start_contigs <- c(
  karyotype$Chr[2],
  karyotype$Chr[3]
)
end_contigs <- c(
  karyotype$Chr[karyotype.length],
  karyotype$Chr[karyotype.length - 1]
)

synteny <- synteny |>
  group_by(Species2) |>
  summarize(
    Species1, Start1, End1, Species2, Start2, End2, chromosome, fill,
    # -1 because a Start of 1 means no offset
    Start2.offset = if_else(Species2 %in% start_contigs, 
                            min(Start2) - 1, 
                            0)) |>
  mutate(
    Start2 = Start2 - Start2.offset,
    End2 = End2 - Start2.offset) |>
  ungroup()

karyotype <- karyotype |>
  left_join(synteny |> select(Species2, Start2.offset, End2), by = c(
    "Chr" = "Species2"
  )) |>
  group_by(Chr) |>
  summarize(
    Start,
    End = if_else(Chr %in% end_contigs, max(End2), if_else(Chr != proto_x_m5, End - Start2.offset, End)),
    fill, species, size, color) |>
  distinct() |>
  ungroup() |>
  arrange(factor(Chr, levels = contigs_of_interest))

synteny <- synteny |>
  # ideogram requires a specific column order
  select(Species1, Start1, End1, Species2, Start2, End2, chromosome, fill)
``` 

### Fill in gaps

#### Find gaps
```{r}
# Fake contigs to mark start of synteny.
x_start <- "X start"
y_start <- "Y start"

synteny.limits <- synteny |>
  mutate(contig = Species2) |>
  group_by(contig) |>
  summarize(
    start = Start1 |> min(),
    end = End1 |> max(),
    chromosome
  ) |> 
  distinct() |>
  ungroup() |>
  rbind(tibble(
    start = c(0, 0, proto_x_m5.length + 1, proto_x_m5.length + 1),
    end = c(0, 0, proto_x_m5.length + 1, proto_x_m5.length + 1),
    chromosome = c("X", "Y", "X", "Y"),
    contig = c(x_start, y_start, "X end", "Y end")
  )) |>
  arrange(chromosome, start)

synteny.limits

synteny.limits.right <- synteny.limits |>
  summarize(right.start = start, chromosome) |>
  arrange(chromosome, right.start) |>
  select(-chromosome)

index.last_y <- which(synteny.limits$chromosome == "Y") |> last()
index.last_x <- which(synteny.limits$chromosome == "X") |> last()

gaps <- synteny.limits |> 
  slice(-index.last_x, -index.last_y) |>
  cbind(synteny.limits.right |>
         slice(-1, -(index.last_x + 1))) |>
  summarize(
    start = end + 1,
    end = right.start - 1,
    chromosome,
    after = contig) |>
  ungroup() |>
  mutate(contig = " " |> strrep(row_number()))

gaps
```


#### Insert gaps into contig order
```{r}
contigs_of_interest <- c(x_start, y_start) |> 
  append(contigs_of_interest)

insertion_indices <- gaps |>
  mutate(index = after |> purrr::map_dbl(\(contig) which(contigs_of_interest == contig)[[1]])) |>
  select(contig, index) |>
  arrange(index |> desc())

for (i in 1:nrow(insertion_indices)) {
  contig <- insertion_indices$contig[i]
  index <- insertion_indices$index[i]
  contigs_of_interest <- contigs_of_interest |>
    append(contig, after = index)
}
# Positions 1, 3: fake marker contigs for the start
contigs_of_interest <- contigs_of_interest[-1][-2]
```

#### Insert gaps into karyotype
```{r}
background <- "fcf4dc"

gaps.karyotype <- gaps |>
  summarize(
    Chr = contig,
    Start = 1,
    End = end - start + 1,
    species = "target",
    fill = background,
    size = 0,
    color = background)
karyotype <- karyotype |>
  rbind(gaps.karyotype) |>
  arrange(factor(Chr, levels = contigs_of_interest))
```

#### Insert gaps into synteny
```{r}
gaps.synteny <- gaps |>
  summarize(
    Species1 = proto_x_m5,
    Start1 = start,
    End1 = end,
    Species2 = contig,
    Start2 = 1,
    End2 = end - start + 1,
    chromosome = chromosome,
    fill = background
  )
synteny <- synteny |>
  rbind(gaps.synteny)
```

#### Take out start gaps
The gaps at the start make the library buggy and remove the chromosome name.
Skip the following code if you are creating new labels for the images later in a photo editing tool anyways

```{r}
skip <- TRUE
if (!skip) {
  "Removing start gaps"
  contigs_to_remove <- contigs_of_interest[1:2]
  contigs_of_interest <- contigs_of_interest[3:length(contigs_of_interest)]
  synteny <- synteny |>
    filter(!(Species2 %in% contigs_to_remove))
} else {
  "Keeping start gaps"
}
```


### Create And Save Ideogram

#### Functions

```{r}
get_rank <- function(to_rank, ranks) {
  for (i in ranks |> seq_along()) {
    contig <- ranks[[i]]
    to_rank <- replace(to_rank, to_rank == contig, i)
  }
  to_rank |> as.integer()
}

get_filtered_ideogram_data <- function(synteny, chrom) {
  karyotype.with_chromosomes <- karyotype |> 
    filter(species == "ProtoX") |>
    rbind(karyotype |> filter(Chr %in% synteny$Species2))

  karyotype.with_chromosomes <- karyotype.with_chromosomes |>
    mutate(species = if_else(species == "target", 
                             "Proto{chrom} M1" |> glue(), 
                             "ProtoX M5"))
  
  contig_head <- karyotype.with_chromosomes |>
    select(Chr) |>
    slice(1) |>
    unlist()
  
  contig_tail <- karyotype.with_chromosomes |>
    select(Chr) |>
    slice(-1) |>
    unlist()
  
  synteny.ranked <- synteny
  synteny.ranked$Species1 <- synteny.ranked$Species1 |>
    get_rank(contig_head)
  synteny.ranked$Species2 <- synteny.ranked$Species2 |>
    get_rank(contig_tail)
  list(
    karyotype = karyotype.with_chromosomes,
    synteny = synteny.ranked
  )
}

get_ideogram_path <- function(chrom) {
  output_dir <- ensure.output("karyotype")
  "{output_dir}/ideogram_{chrom}" |> glue()
}
create_ideogram <- function(chrom) {
  ideogram_path <- get_ideogram_path(chrom)
  data <- synteny |> 
    filter(chromosome == chrom) |> 
    select(-chromosome) |>
    get_filtered_ideogram_data(chrom)

  ideogram(
    karyotype = data$karyotype,
    synteny = data$synteny |> as.data.frame(),
    output = "{ideogram_path}.svg" |> glue())

  convertSVG(
    "{ideogram_path}.svg" |> glue(),
    device = "png", 
    file = "{ideogram_path}.png" |> glue(), 
    dpi = 500)
}
```

#### Creation

```{r}
create_ideogram("X")
create_ideogram("Y")
```

#### Plotting

```{r}
show_ideogram <- function(chrom) {
  ideogram_path <- get_ideogram_path(chrom)
  ggdraw() +
    draw_image("{ideogram_path}.png" |> glue())
}
#show_ideogram("X")
#show_ideogram("Y")
```

### Show chromosome sizes
```{r}
karyotype |> 
  left_join(synteny |> select(Species2, chromosome) , by = c(
    "Chr" = "Species2"
  )) |> 
  distinct() |>
  mutate(chromosome = if_else(is.na(chromosome), "Proto X [MV]", chromosome)) |>
  mutate(chromosome = if_else(chromosome == "X", "Proto X [MI]", chromosome)) |>
  mutate(chromosome = if_else(chromosome == "Y", "Proto Y [MI]", chromosome)) |>
  mutate(gap = Chr |> trimws() |> nchar() != 0) |>
  # Offset of + 1 because `Start is minimally 1
  mutate(length.with_gaps = End - Start + 1) |>
  mutate(length.only_contigs = length.with_gaps * gap) |>
  group_by(chromosome) |>
  summarize(
    length.mbs.with_gaps = sum(length.with_gaps) / 10^6,
    length.mbs.only_contigs = sum(length.only_contigs) / 10^6) |>
  distinct()
```



